proposals/dark_mode_extension_icons.md

```diff
@@ -0,0 +1,351 @@
+* Cons: Two new keys on setIcon() isn't necessarily better than one new key.
+
+---
+
+### Behavior
+
+Existing manifest key. The behavior of the icons manifest key will remain unchanged.
+```
+{
+  "icons": {
+    "64": "icon_64.png"
+  },
+  "action": {
+    "default_icon": {
+      "64": "action_64.png"
+    }
+  }
+}
+```
+
+Order of precedence. The new `icon_variants` keys and subkeys will take
+precedent, followed by the incumbent `icons` key.
+
+The `dark` icon variant is used when a dark color scheme is used for the
+surrounding UI, often referred to as "dark mode".
+The `light` icon variant is used otherwise.
+
+### New Permissions
+
+N/A.
+
+### Manifest File Changes
+
+#### Optional or required?
+`icon_variants` is an optional key of the top level manifest dictionary.
+`icon_variants` is also an optional sub-key in the action key dictionary.
+
+#### Fallback
+* `icon_variants` will be used if they're supplied. Otherwise `default_icon`
+will be used if it's supplied. If neither are supplied, then `icons` will be
+used.
+
+* icon_variants or variants in action.setIcon() must provide both dark and light
+icon, if they occur.
+
+* Order of precedence for action:
+<api>.action.setIcon() -> action.icon_variants -> action.default_icon ->
+icon_variants -> icons
+
+* Order of precedence for non-action:
+icon_variants -> icons
+
+### FAQ
+* What happens if you specify icon_variants.dark but not icon_variants.light,
+and also icons?
+
+If either is provided, both are expected.
+
+* Do we use the dark mode icon from icon_variants for light as well, or fallback
+to icons?
+
+Both are expected if either "light" or "dark" are supplied.
+
+* What happens if you specify icon_variants.dark but not icon_variants.light, and
+you don't have icons?
+
+This could be a warning and a silent fallback to "icons", as the goal is to try
+to reduce errors as much as possible. Another option could be to fall back to
+the dark mode icon if in light mode and no light mode icon is supplied.
+
+* Do we use the dark mode icon from icon_variants or show an auto-generated icon?
+
+The dark mode icon from variants will be shown if supplied and the user is
+currently in dark mode.
+
+#### Warning or error?
+The new `icon_variants` key (or sub-key) will not cause an error in the event
+that it is invalid. Instead, either ignore it altogether, or emit a warning.
+Warnings are preferred over errors as they're future-resilient.
+
+## Security and Privacy
+
+### Exposed Sensitive Data
+
+N/A.
+
+### Abuse Mitigations
+
+N/A.
+
+### Additional Security Considerations
+
+N/A.
+
+## Alternatives
+
+### Existing Workarounds
+
+1. A developer could ask the user to specify what mode they're in and then
+dynamically set the icon to a dark mode icon using action.setIcon(). That
+wouldn't change the management page icon.
+2. A developer could change icons dynamically to dark mode icons if this is
+true: `window.matchMedia('(prefers-color-scheme: dark)')`.
```

Comment:
This proposal focuses on matching the Dark/Light theme of the **Browser UI** (not OS UI). Note that the browser theme can be set independently of the operating system theme.

At present, the behavior of `window.matchMedia('(prefers-color-scheme: dark)').matches` is not consistent across browsers.
- Safari doesn't provide browser dark/light theme setting, it follows OS's theme.
- On Firefox, `prefers-color-scheme: dark` follows the browser setting, not the OS setting. IMO, this is the correct behavior.
- On Chrome, the behavior is very weird. See the following table.

| OS Theme | Chrome Theme | Web pages and Extension pages |
| ------------- | ------------- | ------------- |
| Light | Light | false (Light) |
| Light | Dark | false (Light) |
| Dark | Light | true (Dark) |
| Dark | Dark | true (Dark) |

On web pages or extension pages in a Chrome tab, `prefers-color-scheme: dark` follows the OS setting, not the browser setting.

| OS Theme | Chrome Theme | A override new-tab extension page |
| ------------- | ------------- | ------------- |
| Light | Light | false (Light) |
| Light | Dark | still false (Light), but refresh the page, it become true (Dark) |
| Dark | Light | still true (Dark), but refresh the page, it become false (Light)  |
| Dark | Dark | true (Dark) |

On a Chrome New-Tab override extension page, `prefers-color-scheme: dark` follows the Browser setting, not the OS setting, but you need to refresh the page to get the updated browser theme when changing the browser theme.

IMO, on Chrome, the behavior of `prefers-color-scheme: dark` is not correct.

Reply:
I agree this sounds like a bug in Chrome with `prefers-color-scheme`, but it shouldn't impact this proposal. I expect extensions to follow the browser theme for the purposes of this proposal.

Reply:
I could either remove the workarounds or list even more possibilities. However, the workaround section is just meant to give an idea of possible mitigations while waiting for full dark-mode support.

Reply:
@hanguokai Thanks so much for putting this together! Sounds like very odd behaviour in Chrome. Is there already a crbug on this one?

As for this proposal. I believe we should allow browsers to de-couple `prefers-color-scheme: dark` from using `color_scheme: dark` for extension icons. This allows browsers to have light-theme mode in general, while using the darker variant for extension icons when it uses a browser theme which is more dark than light (for example, green like the comment button on GitHub).

Reply:
@carlosjeurissen I reported an issue [334682520](https://issues.chromium.org/issues/334682520) last week which marked as duplication of issue [332328864](https://issues.chromium.org/issues/332328864). Both issues are relatively new and not fixed. I think this is a Chrome internal issue. From a user's perspective, it's pretty obvious.

From a practical point of view, `prefers-color-scheme: dark` is not ideal (such as theme changing problem). And that is exactly what this proposal seeks to address.

Reply:
@hanguokai Not sure it hurts if browsers decide to use a dark-shaded theme while using light mode for web content. Some users actually enjoy this mode. The only odd behaviour seems to be the refreshing on the new tab page.

> From a practical point of view, prefers-color-scheme: dark as a workaround is not ideal 

That is exactly my point. Very much in favour of the `color_scheme` feature in this proposal.

Reply:
> Not sure it hurts if browsers decide to use a dark-shaded theme while using light mode for web content. Some users actually enjoy this mode.

If so, the browser should provide two settings for them separately (not one setting). But that's too complicated, and I don't think browsers will do that.

Reply:
> If so, the browser should provide two settings for them separately (not one setting). But that's too complicated, and I don't think browsers will do that.

This seems to be the status quo. If you select a dark theme from the chrome web store, and your OS is light mode. Web pages will still display in light mode. The same when selecting a light theme from the chrome web store while your OS is in dark mode. It will show webpages in dark mode.

Reply:
@carlosjeurissen I didn't consider extension-themes before, only the (browser built-in) mode setting. Without installing a theme, the mode should change both UI and web pages. Without installing a theme, if the browser want users custom them separately, the browser should provide two settings, one setting for UI and one setting for web pages.

Back to the topic, this proposal is to solve the matching problem between the extension icon and the browser UI. `prefers-color-scheme` is a CSS function, that used for web pages. And @solomonkinard said "the workaround section is just meant to give an idea of possible mitigations while waiting for full dark-mode support.". So it doesn't really matter whether to delete this workaround in the proposal.

Reply:
> And @solomonkinard said "the workaround section is just meant to give an idea of possible mitigations while waiting for full dark-mode support.". So it doesn't really matter whether to delete this workaround in the proposal.

Right, I would say there are two purposes to this section:

- Share possible workarounds to interested developers.
- Give context for what we are trying to make easier (if the workarounds were perfect, maybe we wouldn't need this proposal).

I don't think either of those meaningfully change us wanting to do this, so agreed with Jackie that I don't think we need to worry too much about this section :)

Reply:
@oliverdunk Agreed there is no need to make updates to the section. Was just following up on the conversation about how browsers handle theming and dark mode.